### ISSUE TITLE
Enforce CSP on the content being validated

### ISSUE THREAD

**Issue by shekyan** [validator/validator#207]
It'd be fairly easy to use [Salvation](https://github.com/shapesecurity/salvation) (already being used to check CSP parsing errors) to implement enforcement of CSP on the following resources. Salvation includes helpful methods on parsed policies such as `allowsStyleWithHash(HashAlgorithm algorithm, Base64Value hashValue)`, `allowsUnsafeInlineScript()`, and `allowsPlugin(MediaType mediaType)`.
- external script source is allowed by the policy
  
  ``` html
  <script src="..."></script>
  ```
- external style source is allowed by the policy
  
  ``` html
  <link href="..." rel="stylesheet" type="text/css">
  ```
- inline script is allowed (through `unsafe-inline`, `hash-source` or `nonce-source`)
  
  ``` html
  <script>...</script>
  <script nonce="...">...</script>
  ```
  
  ``` html
  <body onload="...">
  ```
- inline style is allowed (through `unsafe-inline`, `hash-source` or `nonce-source`)
  
  ``` html
  <style>...</style>
  <style nonce="...">...</style>
  ```
  
  ``` html
  <body style="...">
  ```
- images, children (iframes, workers) are allowed by the policy
  
  ``` html
  <img src="...">
  ```

Thoughts? This won't be able to confirm that the web page will never see a CSP violation, given resources may be dynamically loaded and eval protection depends on runtime JavaScript values, but it will still be very helpful. 
cc @michaelficarra

**Comment by sideshowbarker** [comment]
Would love to add this. I guess this would go a bit beyond the typical kind of checks that most users expect a “markup validator” to do (e.g., what the old W3C Markup Validator was limited to checking) but being able to offer things like this at another level of help to Web authors/developers is part of the reason I usually try to call this tool a “checker” rather than a “validator”.

Anyway my highest-level goal with the time I put into development of the checker is just make it a tool that helps Web authors/developers catch unintended mistakes they might otherwise miss. So adding messages about CSP enforcement violations would help.

All that said, it may be a while before I could look into implementing this. So in the mean time if somebody wants to write up a patch or short design document or something I would be happy to have it.

One thing to keep in mind is that this checker does not implement a DOM—it works in a fully streaming fashion and operates off (SAX) parse events. So I think if the CSP is specified in a `meta` element (rather than in the HTTP headers) that would mean it could not report any violations that occurred in document order prior to that `meta`.

But in the case of the CSP being specified in the HTTP headers, we could report the errors regardless of where they occur in the document.

**Comment by shekyan** [comment]
For CSPs delivered through `meta` we can at least enforce whatever is enforceable from the https://html.spec.whatwg.org/#attr-meta-http-equiv-content-security-policy, and notify of the risks having `<link>` or '<script>' before 'meta' containing CSP.

I can give it a try implementing this.

**Comment by sideshowbarker** [comment]
> For CSPs delivered through `meta` we can at least enforce whatever is enforceable from the https://html.spec.whatwg.org/#attr-meta-http-equiv-content-security-policy, and notify of the risks having `<link>` or `<script>` before 'meta' containing CSP.

Yeah, for the checker I could just have it emit an info-level message stating as much.

> I can give it a try implementing this.

Super :+1:

**Comment by sideshowbarker** [comment]
Implemented in https://github.com/validator/validator/commit/182a3a4299c36948ddeec4a907f5fafa68da9d7e, and deployed to https://validator.w3.org/nu/

Please try it, and if you find it’s doing anything wrong or unexpected, please file new/separate issues in the tracker.
